File: problems/Closest Binary Search Tree Value/closestValue.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int ClosestValue(TreeNode root, double target) {
        var stack = new Stack<TreeNode>();
        int? prev = null;

        while (0 < stack.Count || null != root) {
            while (null != root) {
                stack.Push(root);
                root = root.left;
            }

            var curr = stack.Pop();

            if (prev.HasValue && target >= prev.Value && target < curr.val) {
                return Math.Abs(target - prev.Value) < Math.Abs(target - curr.val) ? prev.Value : curr.val;
            } else if (target < curr.val) {
                return curr.val;
            }

            prev = curr.val;
            root = curr.right;
        }

        return prev.Value;
    }
}

File: problems/Closest Binary Search Tree Value/closestValue.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int closestValue(TreeNode root, double target) {
        int result = 0;
        double minDiff = Integer.MAX_VALUE;
        List<Integer> inOrder = new ArrayList();

        fillInOrder(root, inOrder);

        for (int idx = 0; inOrder.size() > idx; ++idx) {
            int value = inOrder.get(idx);

            if (Math.abs(target - value) < minDiff) {
                minDiff = Math.abs(target - value);
                result = value;
            }
        }

        return result;
    }

    public void fillInOrder(TreeNode node, List<Integer> inOrder) {
        if (null == node) {
            return;
        }

        fillInOrder(node.left, inOrder);
        inOrder.add(node.val);
        fillInOrder(node.right, inOrder);
    }
}
